### PR TITLE
OO version of in-vessel component geometry construction

### LINKED ISSUE #44: First pass to OO for blanket geometry
Convert blanket geometry building code to object oriented with minimal redesign.

### PR DESCRIPTION
**Pull request by connoramoreno**
Creates an object-oriented version of in-vessel component (IVC) geometry construction. The new Stellarator class features three primary child classes:

- Data class: contains all data for IVCs, including outer surface Cartesian point-loci
- CADGeometry class: contains all CAD information for IVCs, including the corresponding volume IDs in Cubit-generated H5M file. This class also contains the magnet geometry child class MagnetSet
- SourceMesh class: contains the source mesh H5M file

Passes all included tests.

Fixes #44.

### CODE REVIEW

**Comment by gonuke** [comment]
High level design questions:
* I think it makes sense to have this build only the invessel components and completely decouple the magnet and source mesh generation.  
* If so, then I think we also move all the export to somewhere else.

The final design is: 
* parastell
    * reads input
    * initializes cad engines as required, esp cubit
    * calls for each requested component to do its work
        *  ivc
        * magnets
        * source mesh
    * exports data as requested

**Comment by gonuke** [comment]
The follow up question is whether we do that before merging this or after?

**Comment by connoramoreno** [comment]
I think that makes sense! I don't think it would take long to restructure the script to accommodate that.

**Comment by connoramoreno** [comment]
I've separated building the IVCs from the main script as suggested. It's worth noting though that `pytest` fails even though the scripts work. Specifically, running the test scripts using Python doesn't result in any errors and everything runs smoothly. When using `pytest` however, there's a library linking issue related to Cubit:

```
___________________________________________________________________________________________________ ERROR collecting test_parastell.py ____________________________________________________________________________________________________
ImportError while importing test module '/home/camoreno/parastell/tests/test_parastell.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../miniforge3/envs/ps_env/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_parastell.py:1: in <module>
    import src.parastell as ps
../src/parastell.py:1: in <module>
    import cubit
/opt/Coreform-Cubit-2023.11/bin/cubit.py:44: in <module>
    from cubit3 import *
/opt/Coreform-Cubit-2023.11/bin/cubit3.py:12: in <module>
    import _cubit3
E   ImportError: /opt/Coreform-Cubit-2023.11/bin/libshowviz_base.so: undefined symbol: _ZN22vtkXOpenGLRenderWindow11SetWindowIdEm
```

I tried for a while to fix this but I'm really not sure what's going on.

**Review by gonuke — CHANGES_REQUESTED** [review]
I got started with a few comments here, but as I got deeper, I wanted to think about how to simplify this a little.  I put some ideas here: https://docs.google.com/presentation/d/10_hmTwkoqR1gxFOxLZhh6XAQQXSNU8qs8qN-A0s_Htg/edit#slide=id.g2c0036363df_0_9

File: src/ivc.py
Comment (connoramoreno) [reply src/ivc.py:188]: The toroidal extent is also defined independently for the source mesh. I found that keeping the geometry and source toroidal extent definitions independent was useful since one may want to repeat the segment geometry in OpenMC (or other MC code) rather than ParaStell for improved performance. In this case, the source still needs to span the full geometry used in MC, which would be different from the geometry modeled in ParaStell

**Comment by connoramoreno** [comment]
Unfortunately, I realized too late that I forgot to commit renaming the files after doing `git mv` so I lost the git history from ivc.py and test_ivc.py. Luckily, there wasn't much history to begin with.

**Comment by connoramoreno** [comment]
Also unfortunately, the same issue persists with `pytest` (though I didn't expect that to change with the updated structure).

**Comment by connoramoreno** [comment]
I think this updated structure is much cleaner by the way; the in-vessel component script is far more readable, not to mention significantly shorter. Thanks, @gonuke!

File: src/invessel_build.py
Comment (gonuke) [inline src/invessel_build.py:10]: Eventually, we should define this in one place and reuse it.

File: src/invessel_build.py
Comment (gonuke) [inline src/invessel_build.py:119]: Let's initialize these here

```suggestion
        self.sol_h5m_tag = sol_h5m_tag

        self.Surfaces = []
        self.Components = []

```

File: src/invessel_build.py
Comment (gonuke) [inline src/invessel_build.py:165]: Delete this if it is initialized in the constructor above

File: src/invessel_build.py
Comment (gonuke) [inline src/invessel_build.py:187]: I think the above logic of the defaults means that the `h5m_tag` for plasma and sol could be `Null`. Is that OK? Or do we need to test inputs to prevent that.

File: src/invessel_build.py
Comment (gonuke) [inline src/invessel_build.py:235]: Remove this if initialized above

File: src/parastell.py
Comment (gonuke) [inline src/parastell.py:238]: Maybe call this `cubit_plugin_dir` for clarity

File: src/parastell.py
Comment (gonuke) [inline src/parastell.py:265]: Because of how you've imported `source_mesh` doesn't this need to be
```suggestion
        self.source_mesh = source_mesh.SourceMesh(self.vmec, self.source)
```

File: src/parastell.py
Comment (gonuke) [inline src/parastell.py:257]: You import `invessel_build` as `ivb` so I think you are overwriting this variable here.  Instead, let's set it to be the member variable right away. 

```suggestion
        self.invessel_build = InVesselBuild(
            self.vmec, self.build, self.repeat, self.num_phi, self.num_theta,
            self.scale, self.export_dict['plasma_h5m_tag'],
            self.export_dict['sol_h5m_tag'], self.logger
        )
```

File: src/parastell.py
Comment (gonuke) [inline src/parastell.py:520]: I don't think we need this class, right?

File: src/parastell.py
Comment (gonuke) [inline src/parastell.py:561]: We shouldn't need this class

File: src/parastell.py
Comment (gonuke) [inline src/parastell.py:584]: We shouldn't need this class

File: src/parastell.py
Comment (gonuke) [inline src/parastell.py:270]:
```suggestion
        self.magnet_set = magnet_coils.MagnetSet(
```

**Review by gonuke — CHANGES_REQUESTED** [review]
This does look a lot cleaner (if I do say so myself).  A few more comments and suggestions here.  Probably will need one more round of review after I see how these changes settle in.

File: src/invessel_build.py
Comment (connoramoreno) [reply src/invessel_build.py:416]: I've created a utility script (`utils.py`) to house these kinds of utility functions. So far, I've put `normalize()` and `expand_ang()` in the script, but we can discuss this structure.

File: src/invessel_build.py
Comment (connoramoreno) [reply src/invessel_build.py:187]: Good point. The intended workflow is that if `plasma_h5m_tag` and `sol_h5m_tag` are `None`, the applied material tag for each will be set to the respective name (i.e., the tag for the plasma will be `'plasma'` and that for the SOL will be `'sol'`) but it seems I accidentally dropped that functionality during my change to OO. I'll rework the script to make sure that functionality is restored, unless you have a suggestion for an alternative workflow.

**Review by gonuke — CHANGES_REQUESTED** [review]
A couple of final little cleanup/changes

**Comment by gonuke** [comment]
Since I just merged #41 that has a copy of `src/utils.py` this probably needs a rebase

File: src/invessel_build.py
Comment (connoramoreno) [reply src/invessel_build.py:124]: Yes, my bad

**Review by gonuke — APPROVED** [review]
A few more changes... I think the export part is difficult to follow and a little scattered.  I'll be proposing some new design for that, probably.

**Review by gonuke — CHANGES_REQUESTED** [review]
I meant to request changes and not approve the last review.

File: src/sourcemesh.py
Comment (connoramoreno) [reply src/sourcemesh.py:311]: Yes, definitely more readable as it was; my change wasn't intentional, just blindly using an auto-formatter

File: src/parastell.py
Comment (connoramoreno) [reply src/parastell.py:281]: `strengths` is a member variable of the `SourceMesh` class, so it is available as part of that class object. We could alternatively make `strengths` a member variable of the `Stellarator` class after `SourceMesh` is constructed. As is, `strengths` can be accessed via
``` python
stellarator = parastell.Stellarator(...)
stellarator.construct_source_mesh(...)
strengths = stellarator.source_mesh.strengths
```

File: src/parastell.py
Comment (connoramoreno) [reply src/parastell.py:317]: The method essentially combines the in-vessel component and magnet CAD information into a single dictionary but this data structure definitely isn't necessary for the export procedure. I constructed it this way to make component information (volume IDs and material tags) centrally available for the user after CAD construction, meshing, etc.

I figured if the data is going to be structured this way already, I might as well use it, although I agree it might not be the most streamlined methodology.

We can alternatively have the user access this information from separate `InVesselBuild` and `MagnetSet` class objects on the backend.

**Comment by connoramoreno** [comment]
This PR has become rather large, sorry.

File: src/cubit_utils.py
Comment (gonuke) [inline src/cubit_utils.py:19]: More meaningful name here
```suggestion
    if not cubit_initialized:
```

File: src/cubit_utils.py
Comment (gonuke) [inline src/cubit_utils.py:37]: should we call `init_cubit` in each of these functions?  After the first time, it will basically do nothing, but at least we'll always know it's been initialized.

File: src/cubit_utils.py
Comment (gonuke) [inline src/cubit_utils.py:78]: Separation of concerns: Should we separate mesh generation from mesh export?

File: src/cubit_utils.py
Comment (gonuke) [inline src/cubit_utils.py:88]: I think this could live with the IVC stuff.  I was imagining these utils being largely (entirely) about import/export and not about geometry querying/modification.

Similarly for `merge_layer_surface` but happy to discuss.

File: src/cubit_utils.py
Comment (gonuke) [inline src/cubit_utils.py:221]:
```suggestion
    tol_str = ''

    if faceting_tolerance is not None:
        tol_str += f'faceting_tolerance {faceting_tolerance} '
    if length_tolerance is not None:
        tol_str += f'length_tolerance {length_tolerance} '
    if normal_tolerance is not None:
        tol_str += f'normal_tolerance {normal_tolerance}'
```

File: src/cubit_utils.py
Comment (gonuke) [inline src/cubit_utils.py:226]:
```suggestion
        f'export dagmc "{export_path}" {tol_str} make_watertight'
```

File: src/cubit_utils.py
Comment (gonuke) [inline src/cubit_utils.py:200]: Put this in a little function and share it with the `native` method

**Review by gonuke — CHANGES_REQUESTED** [review]
Some changes to the cubit utils and then I'll think we'll merge and revise after merging as needed

File: src/cubit_utils.py
Comment (connoramoreno) [reply src/cubit_utils.py:88]: Separating functionality the way you suggest makes sense to me, but I wonder how this separation of concerns affects the other utility functions.

E.g., with tetrahedral mesh generation (mentioned above) should just the mesh export live within `cubit_utils.py` while the mesh generation lives within `magnet_coils.py`? Similarly, for DAGMC exports, should imprinting/merging and material tagging live within `parastell.py` while `cubit_utils.py` handles only the export?

The legacy and native DAGMC exports handle material tagging and surface faceting differently, so my concern is that this will lead to many rather small functions specific to export methods. That may be fine but I worry it might be overcomplicating things.

Comment (connoramoreno) [reply src/cubit_utils.py:88]: Also, this routine that skips imprinting depends on STEP files being imported to Cubit, which is called by `parastell.py` and updates component information to include volume IDs for material tagging in Cubit. Because that information belongs to `parastell.py` rather than `invessel_build.py`, I wonder if it might be better to put `merge_layer_surface` and `orient_spline_surfaces` in `parastell.py` rather than `invessel_build.py` as suggested.

File: src/cubit_utils.py
Comment (connoramoreno) [reply src/cubit_utils.py:37]: For the Cubit utility script, I only included `init_cubit()` in `import_step_cubit()` since, presumably, if the export functions are being called, Cubit has already been initialized.

However, `init_cubit()` is included in all functions with Cubit functionality outside `cubit_utils.py` that don't rely on Cubit operations already happening.

**Comment by connoramoreno** [comment]
I'd still like to discuss the implementation of default dictionaries before merging this PR

**Review by gonuke — APPROVED** [review]
Thanks for sticking with this @connoramoreno

**Pull request merged**
